Add minimum/maximum angular resolution criterion for CCSN problems

Inhyeok-Song opened the pull request:
I've added new problem-specific runtime parameters `CCSN_Min_AngRes` and `CCSN_Max_AngRes`.

- `CCSN_Min_AngRes` is the minimum angular resolution in degree.

- `CCSN_Max_AngRes` is the maximum allowed angular resolution in degree.

- `CCSN_Min_AngRes` should be larger than `CCSN_Max_AngRes` (`CCSN_Min_AngRes` >`CCSN_Max_AngRes`).

- We always check and refine `CCSN_Min_AngRes` first, in `Flag_Region_CCSN()` and  `Flag_Lightbulb/CoreCollapse()`, and we ignore `CCSN_Max_AngRes` if the cell is refined with the minimum angular resolution criteria.

- I slightly changed the variable names and coding style according to the upcoming `leakage PR`

- I removed `CCSN_MaxRefine_RadFac` which is basically same with `CCSN_Min_AngRes` but not so intuitive.

Inhyeok-Song committed:
Add minimum & maximum angular resolution criterion for CCSN problems

Inhyeok-Song committed:
correct the inequality

Inhyeok-Song committed:
change comment

Inhyeok-Song committed:
add back CCSN_CC_Rot_Fac accidentaly removed

hfhsieh commented on src/TestProblem/Hydro/CCSN/Init_TestProb_Hydro_CCSN.cpp:
```diff
@@ -65,6 +64,8 @@ static int        CCSN_Eint_Mode;                  // Mode of obtaining internal
        bool       CCSN_Is_PostBounce = false;      // boolean that indicates whether core bounce has occurred
 
+       double     CCSN_Min_AngRes;                 // Minimum angular resolution in degree
+       double     CCSN_Max_AngRes;                 // Maximum angular resolution in degree
```
`Minimum / Maximum` -> `minimum / maximum`

Inhyeok-Song replied:
I have changed.

hfhsieh commented on src/TestProblem/Hydro/CCSN/Init_TestProb_Hydro_CCSN.cpp:
```diff
@@ -274,6 +277,11 @@ void SetParameter()
+// CCSN_Min_AngRes must be larger than CCSN_Min_AngRes
+   if ( CCSN_Min_AngRes > 0.0  &&  CCSN_Max_AngRes > 0.0 )
+      if ( CCSN_Min_AngRes <= CCSN_Max_AngRes )
+         Aux_Error( ERROR_INFO, "%s = %d must be larger than %s = %d  !!\n", "CCSN_Min_AngRes", CCSN_Min_AngRes, "CCSN_Max_AngRes", CCSN_Max_AngRes );
```
- combine the `if` statement
- the format `%d` is incorrect
- check `OPT__FLAG_REGION` is enabled when `CCSN_Min_AngRes` and/or `CCSN_Max_AngRes` are set.
- print warning message if `CCSN_Min_AngRes < 2.0 * CCSN_Max_AngRes` in which `CCSN_Max_AngRes` is not strictly followed.
- The code in my version is not optimized yet. I prefer to convert the unit of `CCSN_Min/Max_AngRes` to radian in this function in order to save time.

Inhyeok-Song replied:
- I combined the `if` statement
- I corrected the format
- I added checking `OPT__FLAG_REGION`
- I also added this warning message
- I converted `CCSN_Ang_Min/Max` to radian

hfhsieh commented on src/TestProblem/Hydro/CCSN/Init_TestProb_Hydro_CCSN.cpp:
```diff
@@ -933,6 +942,9 @@ void Init_TestProb_Hydro_CCSN()
    Flu_ResetByUser_Func_Ptr = Flu_ResetByUser_CCSN;
 #  endif
 
+   if ( CCSN_Prob != 0 )
```
replace `0` with `Migration_Test`

Inhyeok-Song replied:
I have replaced this.

hfhsieh commented on src/TestProblem/Hydro/CCSN/Flag_CCSN.cpp:
```diff
@@ -34,46 +39,44 @@ extern double CCSN_CentralDens;
+// (3) check if the minimum angular resolution (in degree) is reached
+   if ( CCSN_Min_AngRes > 0.0  && !MaxRefine  &&  !Flag )
+      if ( R * CCSN_Min_AngRes < dh * RAD2DEG )
+         Flag = true;
```
- move this into the `if ( !MaxRefine )` block above, and remove unnecessary conditions
- combine the two `if` statement
- use two spaces before and after `&&`

Inhyeok-Song replied:
- I moved and removed unnecessary conditions.
- I combined this.
- I also corrected the space.

hfhsieh replied:
I think we can combine criteria (2) and (3) further:
``` c
if ( !MaxRefine )
{
//   (2) ...
//   (3) ...
}
```

Inhyeok-Song replied:
I combined this part.

hfhsieh commented on src/TestProblem/Hydro/CCSN/Flag_CCSN.cpp:
```diff
@@ -34,46 +39,44 @@ extern double CCSN_CentralDens;
-   const double dy = Center[1] - Pos[1];
-   const double dz = Center[2] - Pos[2];
-   const double r  = sqrt(  SQR( dx ) + SQR( dy ) + SQR( dz )  );
+   const double Center[3] = { 0.5*amr->BoxSize[0], 0.5*amr->BoxSize[1], 0.5*amr->BoxSize[2] };
```
use `amr->BoxCenter` instead of `0.5*amr->BoxSize` for `Center`. check other places.

Inhyeok-Song replied:
I have replaced them.

hfhsieh commented on src/TestProblem/Hydro/CCSN/Flag_CCSN.cpp:
```diff
@@ -107,36 +110,81 @@ bool Flag_Lightbulb( const int i, const int j, const int k, const int lv, const
-      if ( Rho[k][j][i] < Threshold[0] )   return false;
+// (2) density is larger than the threshold in Input__Flag_User
+   else if ( Rho[k][j][i] < Threshold[0] )
+      Flag = false;
```
The usage `CCSN_MaxRefine_RadFac` is similar to `CCSN_Max_AngRes` but not totally the same. If you want to remove `CCSN_MaxRefine_RadFac`, then this criterion on density is no longer needed.

Inhyeok-Song replied:
I simply removed `CCSN_MaxRefine_RadFac` and corresponding density criterion.

hfhsieh commented on src/TestProblem/Hydro/CCSN/Flag_CCSN.cpp:
```diff
@@ -107,36 +110,81 @@ bool Flag_Lightbulb( const int i, const int j, const int k, const int lv, const
+// (3) check if the minimum angular resolution (in degree) is reached
+   if ( CCSN_Min_AngRes > 0.0  &&  !Flag )
+      if ( R * CCSN_Min_AngRes < dh * RAD2DEG )
+         Flag = true;
```
combine the two `if` statements

Inhyeok-Song replied:
I have combined this.

hfhsieh commented on src/TestProblem/Hydro/CCSN/Flag_CCSN.cpp:
```diff
@@ -107,36 +110,81 @@ bool Flag_Lightbulb( const int i, const int j, const int k, const int lv, const
+   {
+      if ( 2.0 * R * CCSN_Max_AngRes > dh * RAD2DEG )   Within = false;
+
+      if ( !Within )   return Within;
```
- combine the statements of `CCSN_Min_AngRes` and `CCSN_Max_AngRes`. We do not need to check `CCSN_Min_AngRes` if `CCSN_Max_AngRes` is disabled.
- remove `if ( !Within )   return Within;` since there are no other criteria below currently.
- Is `(in degree)` necessary? I add that in my tentative code just for reminders

Inhyeok-Song replied:
- I am not so sure about this part, so I left it. Don't we need `CCSN_Min_AngRes` even `CCSN_Max_AngRes` is disabled?
- I have removed this.
- I removed this comment.

hfhsieh replied:
No. `Flag_Region()` is used to check whether the input patch is allowed for refinement. 

If `Within = false`, we will not check if other refinement criteria (Lohner, `Flag_Corecollapse/Lightbulb`, etc) are satisfied. On the other hand, `Within = true` will not affect whether the input patch will be refined or not.

Inhyeok-Song replied:
I modified this part. Please check this.

hfhsieh reviewed (changes requested):
Looks good to me. I have left some comments. Also,
- I prefer `CCSN_AngRes_Min/Max` more than `CCSN_Min/Max_AngRes`
- Check the density criteria related to `CCSN_MaxRefine_RadFac` again

Inhyeok-Song committed:
add more checkings & arrange the code

hfhsieh commented on src/TestProblem/Hydro/CCSN/Flag_CCSN.cpp:
```diff
@@ -125,14 +123,9 @@ bool Flag_Lightbulb( const int i, const int j, const int k, const int lv, const
    if      ( R * UNIT_L < 3e6 )
```
remove unnecessary spaces after `if`

Inhyeok-Song replied:
I removed the space.

hfhsieh commented on src/TestProblem/Hydro/CCSN/Flag_CCSN.cpp:
```diff
@@ -125,14 +123,9 @@ bool Flag_Lightbulb( const int i, const int j, const int k, const int lv, const
-      if ( R * CCSN_Min_AngRes < dh * RAD2DEG )
-         Flag = true;
+// (2) check if the minimum angular resolution is reached
+   if ( CCSN_AngRes_Min > 0.0  &&  R * CCSN_AngRes_Min < dh  &&  !Flag )
```
move the condition `!Flag` to the first one.

Inhyeok-Song replied:
I move the `!Flag` to the first place.

hfhsieh commented on src/TestProblem/Hydro/CCSN/Flag_CCSN.cpp:
```diff
@@ -167,22 +160,17 @@ bool Flag_Region_CCSN( const int i, const int j, const int k, const int lv, cons
-      if ( !Within )   return Within;
-   }
+// check allowed maximum refine level based on angular resolution
+   if ( CCSN_AngRes_Max > 0.0  &&   2.0 * R * CCSN_AngRes_Max > dh )
```
- remove one space after `&&`

Inhyeok-Song replied:
I removed the space.

hfhsieh commented on src/TestProblem/Hydro/CCSN/Init_TestProb_Hydro_CCSN.cpp:
```diff
@@ -277,10 +277,31 @@ void SetParameter()
-         Aux_Error( ERROR_INFO, "%s = %d must be larger than %s = %d  !!\n", "CCSN_Min_AngRes", CCSN_Min_AngRes, "CCSN_Max_AngRes", CCSN_Max_AngRes );
+// check OPT__FLAG_REGION is enabled for CCSN_AngRes_Min/Max and convert degree to radian
+   const double DEG2RAD = M_PI/180.0;
+   if ( CCSN_AngRes_Min > 0.0 ) {
```
add one empty line before `if`

Inhyeok-Song replied:
I added an emtpy line.

hfhsieh commented on src/TestProblem/Hydro/CCSN/Init_TestProb_Hydro_CCSN.cpp:
```diff
@@ -277,10 +277,31 @@ void SetParameter()
+   const double DEG2RAD = M_PI/180.0;
+   if ( CCSN_AngRes_Min > 0.0 ) {
+      CCSN_AngRes_Min *= DEG2RAD;
+      if ( !OPT__FLAG_REGION )
```
add one empty line before `if`

Inhyeok-Song replied:
I added an emtpy line.

hfhsieh commented on src/TestProblem/Hydro/CCSN/Init_TestProb_Hydro_CCSN.cpp:
```diff
@@ -277,10 +277,31 @@ void SetParameter()
+   if ( CCSN_AngRes_Min > 0.0 ) {
+      CCSN_AngRes_Min *= DEG2RAD;
+      if ( !OPT__FLAG_REGION )
+         Aux_Error( ERROR_INFO, "%s is disabled for %s = %13.7e!!\n", "OPT__FLAG_REGION", "CCSN_AngRes_Min", CCSN_AngRes_Min );
```
check if we need this

Inhyeok-Song replied:
We don't need this check for `CCSN_AngRes_Min`. So I removed this part.

hfhsieh commented on src/TestProblem/Hydro/CCSN/Init_TestProb_Hydro_CCSN.cpp:
```diff
@@ -277,10 +277,31 @@ void SetParameter()
+
+   if ( CCSN_AngRes_Max > 0.0 ) {
+      CCSN_AngRes_Max *= DEG2RAD;
+      if ( !OPT__FLAG_REGION )
```
add one empty line before if

Inhyeok-Song replied:
I added an empty line.

hfhsieh commented on src/TestProblem/Hydro/CCSN/Init_TestProb_Hydro_CCSN.cpp:
```diff
@@ -277,10 +277,31 @@ void SetParameter()
+   if ( CCSN_AngRes_Max > 0.0 ) {
+      CCSN_AngRes_Max *= DEG2RAD;
+      if ( !OPT__FLAG_REGION )
+         Aux_Error( ERROR_INFO, "%s is disabled for %s = %13.7e!!\n", "OPT__FLAG_REGION", "CCSN_AngRes_Max", CCSN_AngRes_Max );
```
add spaces before `!!`. check other places.

Inhyeok-Song replied:
I added a space before `!!` and checked other places too.

hfhsieh commented on src/TestProblem/Hydro/CCSN/Init_TestProb_Hydro_CCSN.cpp:
```diff
@@ -277,10 +277,31 @@ void SetParameter()
+// check OPT__FLAG_REGION is enabled for CCSN_AngRes_Min/Max and convert degree to radian
+   const double DEG2RAD = M_PI/180.0;
+   if ( CCSN_AngRes_Min > 0.0 ) {
+      CCSN_AngRes_Min *= DEG2RAD;
```
print information that `CCSN_AngRes_Min/Max` have been reset (`PRINT_WARNING()`).

Inhyeok-Song replied:
I added `PRINT_WARNING()` after unit conversion.

hfhsieh commented on src/TestProblem/Hydro/CCSN/Init_TestProb_Hydro_CCSN.cpp:
```diff
@@ -277,10 +277,31 @@ void SetParameter()
+   }
+
+   if (  ( CCSN_AngRes_Min > 0.0  ||  CCSN_AngRes_Max > 0.0 )  &&  !OPT__FLAG_REGION  )
+      Aux_Error( ERROR_INFO, "%s is disabled for %s = %13.7e, %s = %13.7e!!\n",
```
What is the difference between this one and the two checks above?

Inhyeok-Song replied:
I removed this part.

hfhsieh commented on src/TestProblem/Hydro/CCSN/Init_TestProb_Hydro_CCSN.cpp:
```diff
@@ -277,10 +277,31 @@ void SetParameter()
+      if ( CCSN_AngRes_Min <= CCSN_AngRes_Max )
+         Aux_Error( ERROR_INFO, "%s = %13.7e must be larger than %s = %13.7e  !!\n", "CCSN_AngRes_Min", CCSN_AngRes_Min, "CCSN_AngRes_Max", CCSN_AngRes_Max );
+      if ( CCSN_AngRes_Min < 2.0 * CCSN_AngRes_Max )
+         Aux_Message( stdout, "WARNING : CCSN_Min_AngRes < 2.0 * CCSN_Max_AngRes: CCSN_Max_AngRes might not be strictly followed!!\n" );
```
correct `CCSN_Min/Max_AngRes`

Inhyeok-Song replied:
I corrected this.

hfhsieh replied:
There is still one that hasn't been corrected.

Inhyeok-Song replied:
Sorry for the missing, and thanks for the careful review. I corrected this.

hfhsieh commented on src/TestProblem/Hydro/CCSN/Init_TestProb_Hydro_CCSN.cpp:
```diff
@@ -339,8 +360,8 @@ void SetParameter()
-      Aux_Message( stdout, "  minimum angular resolution                          = %13.7e\n", CCSN_Min_AngRes          );
-      Aux_Message( stdout, "  maximum angular resolution                          = %13.7e\n", CCSN_Max_AngRes          );
+      Aux_Message( stdout, "  minimum angular resolution                          = %13.7e\n", CCSN_AngRes_Min/DEG2RAD  );
+      Aux_Message( stdout, "  maximum angular resolution                          = %13.7e\n", CCSN_AngRes_Max/DEG2RAD  );
```
it is fine to print these values in degrees or radians, but it would be better to add the unit in the description.

Inhyeok-Song replied:
I think it is easier to see degree, so I added unit descriptions

Inhyeok-Song committed:
minor updates

Inhyeok-Song committed:
minor updates

Inhyeok-Song committed:
minor updates

Inhyeok-Song committed:
correct the variable name

hyschive commented on src/TestProblem/Hydro/CCSN/Init_TestProb_Hydro_CCSN.cpp:
```diff
@@ -274,6 +277,31 @@ void SetParameter()
    }
 
+// check OPT__FLAG_REGION is enabled for CCSN_AngRes_Max and convert degree to radian
+   const double DEG2RAD = M_PI/180.0;
```
Change to `Deg2Rad`. Variables in all upper-case letters are for symbolic constants and general-purpose global variables loaded from `Input__Parameter` in the gamer convention.

Inhyeok-Song replied:
I have changed the variable name (`DEG2RAD->Deg2Rad`)

hyschive commented on src/TestProblem/Hydro/CCSN/Flag_CCSN.cpp:
```diff
@@ -107,36 +106,69 @@ bool Flag_Lightbulb( const int i, const int j, const int k, const int lv, const
+
+// check allowed maximum refine level based on angular resolution
+   if (  CCSN_AngRes_Max > 0.0  &&  2.0 * R * CCSN_AngRes_Max >  dh  &&
+      !( CCSN_AngRes_Min > 0.0  &&        R * CCSN_AngRes_Min >= dh )  )
```
[IMPORTANT] Shouldn't it be `R * CCSN_AngRes_Min < dh` instead of `R * CCSN_AngRes_Min >= dh`?

Inhyeok-Song replied:
Yes. You are right. I was confused about this. I have changed the condition.

hyschive commented on example/test_problem/Hydro/CCSN/Input__TestProb.CoreCollapse:
```diff
@@ -39,6 +38,9 @@ CCSN_CC_Rot_Fac          -1.0                      # multiplication factor for t
 CCSN_Is_PostBounce        0                        # boolean that indicates whether core bounce has occurred   [0]
 
+CCSN_AngRes_Min          -1.0                      # minimum angular resolution in degree [-1.0]
+CCSN_AngRes_Max          -1.0                      # maximum angular resolution in degree [-1.0]
```
- Add the comment `(<=0=off)`
- "degree" --> "degrees"
- Apply the above modifications to other `Input__TestProb.*` files.

Inhyeok-Song replied:
- I added the `(<=0=off)` comment.
- I have changed "degree" --> "degrees"
- I also changed other `Input_TestProb.*` accordingly.

hyschive commented on src/TestProblem/Hydro/CCSN/Init_TestProb_Hydro_CCSN.cpp:
```diff
@@ -274,6 +277,31 @@ void SetParameter()
+         Aux_Error( ERROR_INFO, "%s = %13.7e must be larger than %s = %13.7e !!\n", "CCSN_AngRes_Min", CCSN_AngRes_Min, "CCSN_AngRes_Max", CCSN_AngRes_Max );
+
+      if ( MPI_Rank == 0  &&  CCSN_AngRes_Min < 2.0 * CCSN_AngRes_Max )
+         Aux_Message( stdout, "WARNING : CCSN_AngRes_Min < 2.0 * CCSN_AngRes_Max: CCSN_AngRes_Max might not be strictly followed !!\n" );
```
Change the warning message to `"WARNING : CCSN_AngRes_Min (%13.7e) < 2.0 * CCSN_AngRes_Max (%13.7e) --> CCSN_AngRes_Max might not be strictly followed !!\n", CCSN_AngRes_Min, 2.0*CCSN_AngRes_Max`

Inhyeok-Song replied:
I have changed the warning message.

hyschive commented on src/TestProblem/Hydro/CCSN/Init_TestProb_Hydro_CCSN.cpp:
```diff
@@ -332,6 +359,8 @@ void SetParameter()
       if ( CCSN_CC_Rot == 2 )
       Aux_Message( stdout, "  multiplication factor for rotational profile        = %13.7e\n", CCSN_CC_Rot_Fac          );
+      Aux_Message( stdout, "  minimum angular resolution (in degree)              = %13.7e\n", CCSN_AngRes_Min/DEG2RAD  );
+      Aux_Message( stdout, "  maximum angular resolution (in degree)              = %13.7e\n", CCSN_AngRes_Max/DEG2RAD  );
```
degree"s"

Inhyeok-Song replied:
I have changed this.

hyschive commented on src/TestProblem/Hydro/CCSN/Init_TestProb_Hydro_CCSN.cpp:
```diff
@@ -933,6 +962,9 @@ void Init_TestProb_Hydro_CCSN()
    Flu_ResetByUser_Func_Ptr = Flu_ResetByUser_CCSN;
 #  endif
 
+   if ( CCSN_Prob != Migration_Test )
```
This hard-coded condition is a bit confusing. Why is it necessary? I think disabling `CCSN_AngRes_Max` by default for the migration test should be sufficient.

hfhsieh replied:
In the PR for the leakage scheme, there is another criterion based on the shock radius in `Flag_Region_CCSN()`. However, we do not measure the shock in the migration test.

hyschive replied:
@hfhsieh Why not just disable `OPT__FLAG_REGION` (and raise an error if it's on) for the migration test?

hfhsieh replied:
`OPT__FLAG_REGION` is already disabled in `Input__Parameter.MigrationTest`. I agree we can remove this condition and raise an error instead.

Inhyeok-Song replied:
Following the above discussion, I have removed the condition (`if ( CCSN_Prob != Migration_Test )`
and add the error condition for `Migration_Test && OPT__FLAG_REGION`

hyschive commented on src/TestProblem/Hydro/CCSN/Flag_CCSN.cpp:
```diff
@@ -34,43 +36,40 @@ extern double CCSN_CentralDens;
 
-//    (2-b) refine the region with r < 30 km
-      if ( r * UNIT_L < 3e6 )
+//    (3-b) refine the region with r < 30 km
```
"with" --> "within"

Inhyeok-Song replied:
I have updated the comment

hyschive commented on src/TestProblem/Hydro/CCSN/Flag_CCSN.cpp:
```diff
@@ -34,43 +36,40 @@ extern double CCSN_CentralDens;
-//    (2-b) refine the region with r < 30 km
-      if ( r * UNIT_L < 3e6 )
+//    (3-b) refine the region with r < 30 km
+      if ( R * UNIT_L < 3e6 )
```
How about making `3e6` a runtime parameter? If you agree, modify `Flag_Lightbulb()` as well.

Inhyeok-Song replied:
I think it would be good. I made it a new parameter (`CCSN_MaxRefine_Rad`)

hyschive commented on src/TestProblem/Hydro/CCSN/Flag_CCSN.cpp:
```diff
@@ -34,43 +36,40 @@ extern double CCSN_CentralDens;
+      if ( CCSN_AngRes_Min > 0.0  &&  R * CCSN_AngRes_Min < dh )
+         Flag = true;
+
+//    (3) always refined to highest level in the region with r < 30 km
```
- Change the comment to "always refine to the highest level in the region within r < 30 km"
- Update `Flag_Lightbulb()` as well

Inhyeok-Song replied:
- I have changed the comment.
- Also updated.

hyschive commented on src/TestProblem/Hydro/CCSN/Flag_CCSN.cpp:
```diff
@@ -107,36 +106,69 @@ bool Flag_Lightbulb( const int i, const int j, const int k, const int lv, const
-      Flag = ( 0.5 * dh ) > Min_CellWidth;
-   }
+// (2) check if the minimum angular resolution is reached
+   if ( !Flag  &&  CCSN_AngRes_Min > 0.0  &&  R * CCSN_AngRes_Min < dh )
```
Remove `!Flag` for simplicity and matching `Flag_CoreCollapse()`

Inhyeok-Song replied:
I have removed it.

hyschive commented on src/TestProblem/Hydro/CCSN/Flag_CCSN.cpp:
```diff
@@ -107,36 +106,69 @@ bool Flag_Lightbulb( const int i, const int j, const int k, const int lv, const
-   if ( r * UNIT_L < 3e6 )
+   if ( R * UNIT_L < 3e6 )
       Flag = true;
 
```
Should we also "always refine the innermost cells" like `Flag_CoreCollapse()`?

Inhyeok-Song replied:
This might not be unnecessary for finer-resolution cases, but I think it is still safer to have this.

hyschive replied:
Do you plan to add it?

Inhyeok-Song replied:
I am sorry I missed this. I have updated this for `Flag_Lightbulb()`

hyschive commented on src/TestProblem/Hydro/CCSN/Flag_CCSN.cpp:
```diff
@@ -107,36 +106,69 @@ bool Flag_Lightbulb( const int i, const int j, const int k, const int lv, const
 
-   else
-   {
-//    (2-a) density is larger than the threshold in Input__Flag_User
-      if ( Rho[k][j][i] < Threshold[0] )   return false;
```
Then should we disable `OPT__FLAG_USER` in `Input__Parameter.*` and remove `Input__Flag_User.*`?

Inhyeok-Song replied:
I am also thinking about this. I have updated `Input__Parameter.*` accordingly and removed `Input__Flag_User.*`.

hyschive commented on src/TestProblem/Hydro/CCSN/Flag_CCSN.cpp:
```diff
@@ -107,36 +106,69 @@ bool Flag_Lightbulb( const int i, const int j, const int k, const int lv, const
+   const double R         = sqrt( SQR(dR[0]) + SQR(dR[1]) + SQR(dR[2]) );
+
+
+// check allowed maximum refine level based on angular resolution
```
Change to "check the maximum allowed refinement level based on angular resolution"

Inhyeok-Song replied:
I have changed the comment.

hyschive reviewed (changes requested):
@Inhyeok-Song Thanks for the contribution! I've added some comments. The only relevant one is highlighted by `[IMPORTANT]` in `Flag_Region_CCSN()`.

Inhyeok-Song committed:
fix the angular resolurion condition and add 'CCSN_MaxRefine_Rad'

hyschive commented on example/test_problem/Hydro/CCSN/Input__TestProb.CoreCollapse:
```diff
@@ -38,8 +38,9 @@ CCSN_CC_Rot_Fac          -1.0                      # multiplication factor for t
 
-CCSN_AngRes_Min          -1.0                      # minimum angular resolution in degree [-1.0]
-CCSN_AngRes_Max          -1.0                      # maximum angular resolution in degree [-1.0]
+CCSN_MaxRefine_Rad        3.0e6                    # radius within to refine to the allowed maximum level in km [3.0e6]
```
- Change the comment to "radius in km within which to refine to the maximum allowed level [3.0e6]"
- Update other `Input__TestProb.*` files.

Inhyeok-Song replied:
I have changed the comments

hyschive commented on src/TestProblem/Hydro/CCSN/Init_TestProb_Hydro_CCSN.cpp:
```diff
@@ -326,41 +332,42 @@ void SetParameter()
-      Aux_Message( stdout, "  minimum angular resolution (in degree)              = %13.7e\n", CCSN_AngRes_Min/DEG2RAD  );
-      Aux_Message( stdout, "  maximum angular resolution (in degree)              = %13.7e\n", CCSN_AngRes_Max/DEG2RAD  );
+      Aux_Message( stdout, "  multiplication factor for rotational profile               = %13.7e\n", CCSN_CC_Rot_Fac          );
+      Aux_Message( stdout, "  radius within to refine to the allowed maximum level in km = %13.7e\n", CCSN_MaxRefine_Rad       );
```
Change to "radius within which to refine to the maximum allowed level (in km)

Inhyeok-Song replied:
I have changed the comment.

hyschive commented on src/TestProblem/Hydro/CCSN/Init_TestProb_Hydro_CCSN.cpp:
```diff
@@ -233,6 +235,10 @@ void SetParameter()
 
+// check OPT__FLAG_REGION is disabled for the migration test
+   if ( CCSN_Prob == Migration_Test  &&  OPT__FLAG_REGION )
+      Aux_Error( ERROR_INFO, "%s is enabled for %s = %s !!\n", "OPT__FLAG_REGION", "CCSN_Prob", Migration_Test );
```
Replace "enabled" with "is not supported"

Inhyeok-Song replied:
I have replaced this.

hyschive reviewed (commented):
@Inhyeok-Song I've added some final minor comments.

Besides, for simplicity and clarity, how about raising an _error_ instead of a warning when `CCSN_AngRes_Min < 2.0*CCSN_AngRes_Max`? Then we don't need the second check in the following lines of `Flag_Region_CCSN()`:
```
   if (  CCSN_AngRes_Max > 0.0  &&  2.0 * R * CCSN_AngRes_Max > dh  &&
      !( CCSN_AngRes_Min > 0.0  &&        R * CCSN_AngRes_Min < dh )  )
      Within = false;
```
If you agree, also modify the following lines:
```
      if ( CCSN_AngRes_Min <= CCSN_AngRes_Max )
         Aux_Error( ERROR_INFO, "%s = %13.7e must be larger than %s = %13.7e !!\n", "CCSN_AngRes_Min", CCSN_AngRes_Min, "CCSN_AngRes_Max", CCSN_AngRes_Max );
         
      if ( MPI_Rank == 0  &&  CCSN_AngRes_Min < 2.0 * CCSN_AngRes_Max )
         Aux_Message( stdout, "WARNING : CCSN_AngRes_Min (%13.7e) < 2.0 * CCSN_AngRes_Max (%13.7e) --> CCSN_AngRes_Max might not be strictly followed !!\n", CCSN_AngRes_Min, 2.0*CCSN_AngRes_Max );
```

Inhyeok-Song committed:
minor updates

Inhyeok-Song commented:
I agree with the above [suggestion](https://github.com/hyschive/gamer-fork/pull/117#pullrequestreview-2001070669).

I also updated this part.

hfhsieh commented:
I prefer to keep this feature of `CCSN_AngRes_Min < 2.0 * CCSN_AngRes_Max` since it might be useful for cells in the outer region. For example, is `Dx ~ 31.25 km` sufficient for cells around 1000 km (CCSN_AngRes_Min = 2 degrees)?

However, I do not use `CCSN_AngRes_Min` currently. It is up to @Inhyeok-Song

hyschive commented:
> I prefer to keep this feature of `CCSN_AngRes_Min < 2.0 * CCSN_AngRes_Max` since it might be useful for cells in the outer region. For example, is `Dx ~ 31.25 km` sufficient for cells around 1000 km (CCSN_AngRes_Min = 2 degrees)?

@hfhsieh My logic is that one can always decrease `CCSN_AngRes_Max` when `CCSN_AngRes_Min < 2.0 * CCSN_AngRes_Max` occurs, which will improve instead of decreasing accuracy. What do you think?

hfhsieh commented:
@hyschive Agreed. We can release this restriction later if needed.

hyschive reviewed (commented):
@hfhsieh @Inhyeok-Song In `Flag_CCSN.cpp`, should we use `GREP_Center[]` instead of `amr->BoxCenter[]` in case the center drits away after long-term evolution due to the accumulation of numerical errors?

hfhsieh commented:
@hyschive I think using `GREP_Center[]` is better. In addition, we also need to update that in `Flag_CoreCollapse()` (optional) and `Flag_Lightbulb()` for consistency.

@Inhyeok-Song If you plan to replace `BoxCenter[]` with `GREP_Center[]`, note that `GREP_Center[]` is declared only when `GRAVITY` is enabled

Inhyeok-Song committed:
change the center for Flag_*() to GREP_Center

Inhyeok-Song commented:
I replaced the `BoxCenter[]` to `GREP_Center[]`.

@hfhsieh I think we always check if gravity is enabled in `CCSN problem` Do we need to have macro for this?

hyschive commented:
@Inhyeok-Song The code cannot compile with `--gravity=False`.

hfhsieh commented:
@Inhyeok-Song Yes. The sanity check in `Validate()` aims to ensure `GRAVITY` is enabled before performing CCSN simulations, and is not for compilation.

Also, I am unsure whether `GREP_Center[]` is defined before mesh refinements in the initialization stage. Remember to check that carefully.

hyschive commented:
> I am unsure whether GREP_Center[] is defined before mesh refinements in the initialization stage.

I don't think it's defined. You can hard-code `GREP_Center[]` to the box center in `Init_ResetParameter()`.

Inhyeok-Song committed:
add GRAVITY macro and add dummy Input__Flag_User.* back

Inhyeok-Song committed:
remove useless output

Inhyeok-Song committed:
initialized GREP_Center

Inhyeok-Song commented:
@hyschive @hfhsieh 
Thanks for your careful reviews.

- I added the checking `GRAVITY` for `GREP_Center`.

- I added back `Input__Flag_User`. We need to turn on `OPT__FLAG_USER` and it requires `Input__Flag_User` regardless of whether we use it. So I added dummies `Input__Flag_User.*` and changed the parameters accordingly. If you have a better idea, please let me know.

- I added the initialization of `GREP_Center` in `Init_ResetParameter()`, but I am not so sure this is what you intended. If you think this is a wrong implementation, let me know and give me some ideas.

hfhsieh commented on example/test_problem/Hydro/CCSN/Input__Flag_User.CoreCollapse:
```diff
@@ -1,10 +1,10 @@
+      5                    0.0 
+      6                    0.0 
+      7                    0.0 
+      8                    0.0 
```
Remove the trailing spaces in `Input__Flag_User.*`, but keep the empty line at the end of files.

Inhyeok-Song replied:
I have removed the spaces.

hfhsieh commented on src/Init/Init_ResetParameter.cpp:
```diff
@@ -1048,6 +1048,12 @@ void Init_ResetParameter()
 #  ifdef GRAVITY
+// initialize GREP center to the box center
+   extern double GREP_Center[3];
+   GREP_Center[0] = amr->BoxCenter[0];
```
add one empty line

Inhyeok-Song replied:
I added a emtpy line.

hfhsieh commented on src/TestProblem/Hydro/CCSN/Flag_CCSN.cpp:
```diff
@@ -34,43 +41,43 @@ extern double CCSN_CentralDens;
+#  ifdef GRAVITY
+   const double dR [3] = { Pos[0]-GREP_Center[0], Pos[1]-GREP_Center[1], Pos[2]-GREP_Center[2] };
+#  else
+   const double dR [3] = { NULL };
```
I prefer to define `Center = amr->BoxCenter` or `GREP_Center` according to `GRAVITY`. If you want to modify `dR` instead, set it to `Pos - amr->BoxCenter` when `GRAVITY` is disabled.

Inhyeok-Song replied:
I have updated this.

hyschive commented:
@Inhyeok-Song @hfhsieh One issue is that we don't know the correct `GREP_Center[]` during restart when using the potential minimum to determine the center. This could be a serious problem if `GREP_Center[]` has drifted away from the box center, which will affect both the GREP calculation and grid refinement after restart.

I think there are at least two possible solutions.
1. Always store the potential data on the disk (see gamer-project/gamer#298).
2. Store `GREP_Center[]` in HDF5 snapshots, similar to how `AveDensity_Init` is handled in `Output_DumpData_Total_HDF5.cpp` and `Init_ByRestart_HDF5.cpp`.

Actually, we may want to support both methods since each of them has pros and cons. What do you think?

Inhyeok-Song commented:
@hyschive 
I agree with you. This might be dangerous in the long term.
Do you think we should implement this feature in this PR? or is it better to separate it?

Inhyeok-Song committed:
minor change in coding style

hyschive commented:
We can incorporate the second method in this PR.

hfhsieh commented:
@Inhyeok-Song @hyschive 
I prefer the second method because the potential data is not necessary if users adopt other definitions of the GREP center.

Also, I think it should be included in this PR since it is correlated.

hyschive commented:
@hfhsieh Agree. But I'll still implement the first method in `main` since we often need the potential data.

@Inhyeok-Song Let me know if you encounter any problems when implementing the second method. Please test it carefully and confirm with the tool `gamer_compare_data` that the results after restart are bitwise identical.

hyschive commented:
@Inhyeok-Song Any update?

Inhyeok-Song commented:
@hyschive 
Not yet. I am still implementing this.

Inhyeok-Song committed:
record GREP center in HDF5 file to track, in case the center is drifted

Inhyeok-Song commented:
@hyschive @hfhsieh 

I have updated the feature, recording `GREP_Center` for restart runs.

- `GREP_Center` is recorded in HDF5 format.
- We load `GREP_Center` for restart runs.

I ran a fresh run with `core-collapse` and another restart run from the 5th snapshot.
I checked the 10th snapshots from each run (fresh run, restart run), if they are bitwise identical.
And they are identical.

Since the `GREP_Center` is always at the center,
just for the test purpose I commented out the below condition. 
https://github.com/hyschive/gamer-fork/blob/e850e2ebfdd7480a59708461cbf99eec5a0117f9/src/SelfGravity/Poi_UserWorkBeforePoisson_GREP.cpp#L110-L121
They are identical.

If you have time, please review this.

hyschive commented:
@hfhsieh Could you help review it again? I'll take another look after you approve it. Thanks!

hfhsieh commented:
@Inhyeok-Song In the collapse stage, the GREP center should be the box center, and thus your test is not quite meaningful to me. Could you use the box center and add some artificial drifts (like `BoxCenter - 1.0 * Step`) to test if the GREP center is recorded and loaded correctly?

Inhyeok-Song commented:
@hfhsieh 
For the test-purpose I commented out below

       if (  GREP_CENTER_METHOD == GREP_CENTER_DENS  || 
             GREP_CENTER_METHOD == GREP_CENTER_POT     ) 
       { 
          const double Extrema_dh = amr->dh[ Extrema.Level ]; 
  
          if (  fabs( GREP_Center[0] - amr->BoxCenter[0] ) < Extrema_dh  && 
                fabs( GREP_Center[1] - amr->BoxCenter[1] ) < Extrema_dh  && 
                fabs( GREP_Center[2] - amr->BoxCenter[2] ) < Extrema_dh    ) 
             for (int i=0; i<3; i++)   GREP_Center[i] = amr->BoxCenter[i]; 
       } 
    } // if ( lv == 0 ) 
    
Though it is not drifting away far, it is still moving around among the innermost cells.

If you think this is not sufficient, I can test the one you mentioned above.

hfhsieh commented:
@Inhyeok-Song Got it.

hfhsieh commented on src/Auxiliary/Aux_TakeNote.cpp:
```diff
@@ -1253,7 +1253,10 @@ void Aux_TakeNote()
+      fprintf( Note, "GREP_OPT_PRES                   %d\n",      GREP_OPT_PRES           );
+      fprintf( Note, "GREP_Center_X                   %20.14e\n", GREP_Center[0]          );
+      fprintf( Note, "GREP_Center_Y                   %20.14e\n", GREP_Center[1]          );
+      fprintf( Note, "GREP_Center_Z                   %20.14e\n", GREP_Center[2]          ); }
```
`GREP_Center` is not a runtime parameter. Its value could change with time. Why do we need to record them in `Record__Note`?

Inhyeok-Song replied:
Maybe we don't need this. I removed this part.

hfhsieh commented on src/Init/Init_ByRestart_v1.cpp:
```diff
@@ -166,9 +166,10 @@ void Init_ByRestart_v1( const char FileName[] )
+   if ( FormatVersion >= 1200 ) {
 #  ifdef GRAVITY
    fread( &AveDensity_Init, sizeof(double),           1, File );
+   fread( GREP_Center,      sizeof(double),           3, File ); }
```
Syntax error if `GRAVITY` is disabled. Move the `}` after `#endif`, or add one more to `fseek()`.

Inhyeok-Song replied:
I movee the } after `#endif`

hfhsieh commented on src/Output/Output_DumpData_Total.cpp:
```diff
@@ -712,6 +712,7 @@ void Output_DumpData_Total( const char *FileName )
 
 #     ifndef GRAVITY
       const double AveDensity_Init     = NULL_REAL;
+      const double GREP_Center[3]      = { NULL_REAL };
```
This will initialize only the first element to `NULL_REAL`. Have you checked the other two values are `NULL_REAL`?

Inhyeok-Song replied:
You are right. I corrected this.

hfhsieh commented on src/SelfGravity/Poi_UserWorkBeforePoisson_GREP.cpp:
```diff
@@ -25,7 +25,6 @@ Profile_t *GREP_EffPot  [NLEVEL  ][2];
 int    GREP_LvUpdate;
 int    GREP_Sg     [NLEVEL];
 double GREP_SgTime [NLEVEL][2];
-double GREP_Center [3];
```
Remove the declaration of `GREP_Center` in other places since it is now a global variable.

Inhyeok-Song replied:
I removed the declaration of `GREP_Center` in other places.

hfhsieh commented on include/Typedef.h:
```diff
@@ -379,10 +379,11 @@ const GREP_PresScheme_t
+   GREP_CENTER_DENS    = 2,
+   GREP_CENTER_POT     = 3,
+   GREP_CENTER_COM     = 4,
+   GREP_CENTER_RESTART = 5;
```
I prefer to call it `GREP_CENTER_NONE = 0` to indicate that `GREP_Center` is not assigned (e.g., is set when reading the restart file).

Inhyeok-Song replied:
I have changed `GREP_CENTER_RESTART` to `GREP_CENTER_NONE`

hfhsieh commented on src/Init/Init_GAMER.cpp:
```diff
@@ -274,7 +274,14 @@ void Init_GAMER( int *argc, char ***argv )
+            GREP_CENTER_METHOD = GREP_CENTER_RESTART;
+
+         else
+            GREP_CENTER_METHOD = GREP_CENTER_BOX;
```
use the ternary operator here.

Inhyeok-Song replied:
I updated this.

hfhsieh commented on src/SelfGravity/Poi_UserWorkBeforePoisson_GREP.cpp:
```diff
@@ -102,6 +101,9 @@ void Poi_UserWorkBeforePoisson_GREP( const double Time, const int lv )
             Aux_Error( ERROR_INFO, "GREP_CENTER_COM has not been implemented yet!!\n" );
          break;
 
+         case GREP_CENTER_RESTART: // GREP center from RESTART data
```
Update the comment to `restart file`.

Inhyeok-Song replied:
Changed this to `GREP_CENTER_NONE` and moved above.

hfhsieh replied:
Update the comment of `GREP_CENTER_NONE`. In fact, the `GREP_Center` is unchanged in this case, instead of not assigned. Also, it is clear that switch block is for the GREP center, and thus I think it is unnecessary to mention it again in the comment.

Inhyeok-Song replied:
I have just removed the comment.

hfhsieh replied:
So you think that the name `GREP_CENTER_NONE` is clear and does not require any comments? I meant removing the beginning term "GREP center" in the comment, rather than all the comments.

Inhyeok-Song replied:
What exact comment do you want?

hfhsieh replied:
`restart file`

Inhyeok-Song replied:
I updated the comment.

Inhyeok-Song committed:
minor updates

Inhyeok-Song committed:
minor update

Inhyeok-Song committed:
minor update

Inhyeok-Song committed:
change every center to GREP_Center

Inhyeok-Song commented:
I also changed the `GW signal` center and other centers to `GREP_Center`

Inhyeok-Song committed:
correct the unit in comment

hfhsieh commented on src/TestProblem/Hydro/CCSN/Init_TestProb_Hydro_CCSN.cpp:
```diff
@@ -53,7 +53,7 @@ static int        CCSN_Eint_Mode;                  // Mode of obtaining internal
        double     CCSN_CC_MaxRefine_Dens1;         // central density threshold that reduces the maximum refinement level 1
        double     CCSN_CC_MaxRefine_Dens2;         // central density threshold that reduces the maximum refinement level 2
-       double     CCSN_MaxRefine_Rad;              // radius in km within which to refine to the maximum allowed level
+       double     CCSN_MaxRefine_Rad;              // radius in cm within which to refine to the maximum allowed level
```
Also correct the comments in `Input__TestProb.*`.

Inhyeok-Song committed:
correct the unit in comment

hyschive commented on example/test_problem/Hydro/CCSN/Input__Flag_User.CoreCollapse:
```diff
@@ -0,0 +1,10 @@
+# Level                Density
```
Change the header `Density` to, for example, `Dummy`. Same for other `Input__Flag_User.*` files.

Inhyeok-Song replied:
I have updated the headers.

hyschive commented on src/Init/Init_ByRestart_v1.cpp:
Don't change this file, which is no longer used.

Inhyeok-Song replied:
I changed back this file.

hyschive commented on src/Output/Output_DumpData_Total.cpp:
Don't change this file, which is deprecated.

Inhyeok-Song replied:
I changed back this file.

hyschive commented on src/Init/Init_ByRestart_v2.cpp:
Don't change this file, which is deprecated.

Inhyeok-Song replied:
I changed back this file.

hyschive commented on include/HDF5_Typedef.h:
```diff
@@ -74,6 +74,7 @@ struct KeyInfo_t
    double dTime_AllLv[NLEVEL];
 #  ifdef GRAVITY
    double AveDens_Init;             // AveDensity_Init
+   double GREP_Center[3];           // GREP center
```
Also store this variable in the `InputPara_t` structure together with other GREP variables. I know this is a bit confusing and redundant. We may consider removing `KeyInfo_t` completely in the future.

Update `Init_ByRestart_HDF5.cpp` and `Output_DumpData_Total_HDF5.cpp` accordingly.

hyschive replied:
@Inhyeok-Song Upon reflection, I withdraw the above suggestion since `GREP_Center[]` is not an input parameter. Besides, your current implementation is consistent with other existing key variables like `double Time[NLEVEL]`.

hyschive reviewed (commented):
@Inhyeok-Song Looks good. I add some final minor comments.

Inhyeok-Song committed:
roll back the deprecated files

Inhyeok-Song committed:
change the headers density->dummy
